internal/controllers/addon/addon_deletion_reconciler.go

```diff
@@ -52,13 +55,17 @@ func (r *addonDeletionReconciler) Reconcile(ctx context.Context, addon *addonsv1
 	// We set ReadyToBeDeleted=false status condition in response to the delete signal received from OCM.
 	reportAddonReadyToBeDeletedStatus(addon, metav1.ConditionFalse)
 
+	reconErr := metrics.NewReconcileError("addon", r.recorder, true)
+
 	for _, handler := range r.handlers {
 		if err := handler.NotifyAddon(ctx, addon); err != nil {
+			err = reconErr.Join(err, controllers.ErrNotifyAddon)
 			return ctrl.Result{}, err
 		}
 		// If ack is received from the underlying addon, we report ReadyToBeDeleted = true.
 		ackReceived, err := handler.AckReceivedFromAddon(ctx, addon)
 		if err != nil {
+			err = reconErr.Join(err, controllers.ErrAckReceivedFromAddon)
```

Comment:
I see we're appending the error to the metric here, where are we reporting this metric?

Reply:
@supreeth7 So with regards to collecting and reporting,

1. Reconcile errors are collected as metrics from the top-level Reconcile function.
  - This enables all reconcile errors to be collected
  - This avoids duplication of error collection per reconile iteration
  - This also simplifies collection as there will be only few places (done) to do them
  
An observation here is also that Controllers are not designed uniformly, some use subreconcilers, some use/call it phases, some do not use them. Errors are also obviously are not reported specifically. The changes made in this PR made a very generic assumption that;

1. There is a top-level Reconcile function
2. All sub-Reconcile function are handled generally by `subreconcilers` or whatever it's called 
3. Reconcile errors are collected from the top-level Reconcile function, otherwise they are simply returned

Reply:
@supreeth7
> I see we're appending the error to the metric here,

Yes that's right. To be more precise, this is collating the original error plus (new) our definition of it. Previously these errors weren't defined elsewhere.

> where are we reporting this metric?
Great question.

So as I've observed controllers are not made uniformly, some use one-level Reconcilers, some use two-level, and they can be called differently in controlelrs (e.g. subreconcilers, reconciler, phase, etc).

They are collected at the top-level Reconcile function for the following reasons.
1. It avoids duplication of error metrics per reconcile iteration.
2. Given the observation above, it keeps the code simpler by discouraging collection of error metrics anywhere in the reconcile code.
3. All reconcile errors are collected automatically from the top-level Reconcile function, everything else just needs to be reported back (e.g. from subreconcilers).